src/NLPModels.jl

```diff
@@ -13,91 +12,75 @@ with `σ = obj_weight`
+export AbstractNLPModelMeta, NLPModelMeta, AbstractNLPModel, Counters, reset!, sum_counters, push!,
+       varscale, lagscale, conscale, NotImplementedError
+
+# Single objective
+export single_obj, single_grad, single_grad!, single_hess, single_hprod, single_hprod!, single_hess_op, single_hess_op!,
+       single_hess_structure, single_hess_structure!, single_hess_coord, single_hess_coord!
+
+# Nonlinear least-squares
+export residual, residual!, jac_residual, jac_coord_residual, jprod_residual, jprod_residual!, jtprod_residual,
+       jtprod_residual!, jac_op_residual, jac_op_residual!, hess_residual, hess_structure_residual!,
+       hess_structure_residual, hess_coord_residual!, hess_coord_residual, jth_hess_residual,
+       hprod_residual, hprod_residual!, hess_op_residual, hess_op_residual!
+
+# Constraints
+export cons, cons!, jth_con, jth_congrad, jth_congrad!, jth_sparse_congrad, jac_structure, jac_coord!,
+       jac_coord, jac, jprod, jprod!, jtprod, jtprod!, jac_op, jac_op!
+
+# Sum of objectives
+export obj, grad, grad!, objgrad, objgrad!, objcons, objcons!, jth_hprod, jth_hprod!, ghjvprod, ghjvprod!,
+       hess_structure!, hess_structure, hess_coord!, hess_coord, hess, hprod, hprod!, hess_op, hess_op!
 
 # import methods we override
 import Base.push!
 import LinearOperators.reset!
 
 include("nlp_utils.jl")
 include("nlp_types.jl")
-include("NLSModels.jl")
 
 mutable struct NotImplementedError <: Exception
   name :: Union{Symbol,Function,String}
 end
 
 Base.showerror(io::IO, e::NotImplementedError) = print(io, e.name, " not implemented")
 
-# simple default API for retrieving counters
-for counter in fieldnames(Counters)
-  @eval begin
-    """`$($counter)(nlp)`
-
-    Get the number of `$(split("$($counter)", "_")[2])` evaluations.
-    """
-    $counter(nlp :: AbstractNLPModel) = nlp.counters.$counter
-    export $counter
-  end
-end
-
-"""`increment!(nlp, s)`
-
-Increment counter `s` of problem `nlp`.
-"""
-function increment!(nlp :: AbstractNLPModel, s :: Symbol)
-  setfield!(nlp.counters, s, getfield(nlp.counters, s) + 1)
-end
-
-"""`sum_counters(counters)`
-
-Sum all counters of `counters`.
+# Methods to be overridden in other packages.
 """
-sum_counters(c :: Counters) = sum(getfield(c, x) for x in fieldnames(Counters))
-
-"""`sum_counters(nlp)`
+    f = single_obj(nlp, x)
 
-Sum all counters of problem `nlp`.
+Evaluate ``fₛ(x)``, the single objective function of `nlp` at `x`.
 """
-sum_counters(nlp :: AbstractNLPModel) = sum_counters(nlp.counters)
+single_obj(::AbstractNLPModel, ::AbstractVector) =
+  throw(NotImplementedError("single_obj"))
 
-"""`reset!(counters)`
+"""`f = obj(nlp, x)`
 
-Reset evaluation counters
+Evaluate ``f(x)``, the objective function of `nlp` at `x`.
 """
-function reset!(counters :: Counters)
-  for f in fieldnames(Counters)
-    setfield!(counters, f, 0)
+function obj(nlp::AbstractNLPModel, x::AbstractVector)
+  increment!(nlp, :neval_obj)
+  fx = zero(eltype(x))
+  if nlp.meta.hassobj
```

Comment:
`has_single_obj`?

Reply:
Fixed

Reply:
Should be consistent with `general_obj` I suppose. How about `has_general_obj`?

Reply:
Yes, I changed to that.